web/src/components/graph/areagraph.js

```diff
@@ -0,0 +1,132 @@
+import React, {
+  useState,
+  useEffect,
+  useRef,
+  useMemo,
+} from 'react';
+import { scaleTime, scaleLinear } from 'd3-scale';
+import { first, last } from 'lodash';
+import moment from 'moment';
+
+import AreaGraphGradients from './areagraphgradients';
+import AreaGraphLayers from './areagraphlayers';
+import GraphBackground from './graphbackground';
+import GraphHoverLine from './graphhoverline';
+import ValueAxis from './valueaxis';
+import TimeAxis from './timeaxis';
+
+const X_AXIS_HEIGHT = 20;
+const Y_AXIS_WIDTH = 35;
+const Y_AXIS_PADDING = 4;
+
+const getTimeScale = (containerWidth, datetimes, currentTime) => scaleTime()
+  .domain([first(datetimes), currentTime ? moment(currentTime).toDate() : last(datetimes)])
+  .range([0, containerWidth]);
+
+const getValueScale = (containerHeight, maxTotalValue) => scaleLinear()
+  .domain([0, maxTotalValue * 1.1])
+  .range([containerHeight, Y_AXIS_PADDING]);
+
+const AreaGraph = React.memo(({
+  id,
+  layers,
+  currentTime,
+  maxTotalValue,
+  valueAxisLabel,
+  datetimes,
+  fillSelector,
+  mouseMoveHandler,
+  mouseOutHandler,
+  layerMouseMoveHandler,
+  layerMouseOutHandler,
+  selectedTimeIndex,
+  selectedLayerIndex,
+  gradientIdPrefix,
```

Comment:
is there a way we can rid of id/selector for the gradient?

Reply:
Yes, that's a good point! I appended it to `layers` props instead of passing it through these 3 separate props so now gradients are conditionally being rendered together with their corresponding layers and their usage is a bit more robust - see https://github.com/tmrowco/electricitymap-contrib/pull/2171/commits/4a35f3d3093e400d3e7e9c895b4eb93903104ea7.